Lib/test/test_capi/test_exceptions.py

```diff
@@ -68,6 +69,46 @@ def test_exc_info(self):
         else:
             self.assertTrue(False)
 
+    def test_warn_with_stacklevel(self):
+        code = textwrap.dedent('''\
+            import _testcapi
+
+            def foo():
+                _testcapi.function_set_warning()
+
+            foo()  # line 6
+
+
+            foo()  # line 9
+        ''')
+        proc = assert_python_ok("-c", code)
+        warnings = proc.err.splitlines()
+        expected = [
+            b'<string>:6: RuntimeWarning: Testing PyErr_WarnEx',
+            b'  foo()  # line 6',
+            b'<string>:9: RuntimeWarning: Testing PyErr_WarnEx',
+            b'  foo()  # line 9'
+        ]
+        self.assertEqual(len(warnings), 4)
+        self.assertEqual(warnings, expected)
+
+    def test_warn_during_finalization(self):
+        code = textwrap.dedent('''\
+            import _testcapi
+
+            class Foo:
+                def __del__(self):
+                    _testcapi.function_set_warning()
+
+            ref = Foo()
+        ''')
+        proc = assert_python_ok("-c", code)
+        warnings = proc.err.splitlines()
+        # Due to the finalization of the interpreter, the source will be ommited
+        # because the ``warnings`` module cannot be imported at this time
+        self.assertEqual(len(warnings), 1)
+        self.assertEqual(warnings[0], b'<sys>:0: RuntimeWarning: Testing PyErr_WarnEx')
```

Comment:
Wait, does it not print the traceback because it cannot import, or because there is nothing to print for `<sys>`? Make the call of `_testcapi.function_set_warning()` indirect in `__del__` (use the intermediate method).

Reply:
> Wait, does it not print the traceback because it cannot import, or because there is nothing to print for `<sys>`?

I do not understand which traceback you are referring to here. There is no error if we try to create a warning that leads to ``import warnings`` during finalization.

Reply:
Sorry, I meant the source line.